integrations/kube-agent-updater/pkg/maintenance/basichttp.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package maintenance
+
+import (
+	"context"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/gravitational/teleport/integrations/kube-agent-updater/pkg/cache"
+)
+
+const (
+	// maintenancePath is the version discovery endpoint representing if the
+	// target version represents a critical update as defined in
+	// https://github.com/gravitational/teleport/blob/master/rfd/0109-cloud-agent-upgrades.md#version-discovery-endpoint
+	maintenancePath = "critical"
+	httpTimeout     = 10 * time.Second
+	cacheDuration   = time.Minute
+)
+
+// basicHTTPMaintenanceClient retrieves whether the target version represents a
+// critical update from an HTTP endpoint. It should not be invoked immediately
+// and must be wrapped in a cache layer in order to avoid spamming the version
+// server in case of reconciliation errors.
+// use BasicHTTPMaintenanceTrigger if you need to check if an update is critical.
+type basicHTTPMaintenanceClient struct {
+	baseURL *url.URL
+	client  *http.Client
+}
+
+// Get sends an HTTP GET request and returns whether the current target version
+// represents a critical update.
+func (b *basicHTTPMaintenanceClient) Get(ctx context.Context) (bool, error) {
+	versionURL := b.baseURL.JoinPath(maintenancePath)
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, versionURL.String(), nil)
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+	res, err := b.client.Do(req)
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusOK {
```

Comment:
Drain the body even if the status was not ok. For some reason, someone can write something into the maintenance path and the next calls can pick up the buffered connection